NorthstarDLL/logging.cpp

```diff
@@ -12,257 +11,194 @@
+		{
+			spdlog::error("Failed creating log file!");
+			MessageBoxA(
+				0, "Failed creating log file! Make sure the profile directory is writable.", "Northstar Warning", MB_ICONWARNING | MB_OK);
+		}
+	}
 }
 
-// used for printing the output of status
-// clang-format off
-AUTOHOOK(Status_ConMsg, engine.dll + 0x15ABD0,
-void,, (const char* text, ...))
-// clang-format on
+void ExternalConsoleSink::sink_it_(const spdlog::details::log_msg& msg)
 {
-	char formatted[2048];
-	va_list list;
-
-	va_start(list, text);
-	vsprintf_s(formatted, text, list);
-	va_end(list);
-
-	auto endpos = strlen(formatted);
-	if (formatted[endpos - 1] == '\n')
-		formatted[endpos - 1] = '\0'; // cut off repeated newline
-
-	spdlog::info(formatted);
+	throw std::runtime_error("sink_it_ called on SourceConsoleSink with pure log_msg. This is an error!");
 }
 
-// clang-format off
-AUTOHOOK(CClientState_ProcessPrint, engine.dll + 0x1A1530, 
-bool,, (void* thisptr, uintptr_t msg))
-// clang-format on
+void ExternalConsoleSink::custom_sink_it_(const custom_log_msg& msg)
 {
-	char* text = *(char**)(msg + 0x20);
+	spdlog::memory_buf_t formatted;
+	spdlog::sinks::base_sink<std::mutex>::formatter_->format(msg, formatted);
 
-	auto endpos = strlen(text);
-	if (text[endpos - 1] == '\n')
-		text[endpos - 1] = '\0'; // cut off repeated newline
+	std::string out = "";
+	// if ansi colour is turned off, just use WriteConsoleA and return
+	if (!g_bSpdLog_UseAnsiColor)
+	{
+		out += fmt::to_string(formatted);
+	}
 
-	spdlog::info(text);
-	return true;
-}
+	// print to the console with colours
+	else
+	{
+		// get message string
+		std::string str = fmt::to_string(formatted);
 
-ConVar* Cvar_cl_showtextmsg;
+		std::string levelColor = m_LogColours[msg.level];
+		std::string name {msg.logger_name.begin(), msg.logger_name.end()};
 
-class ICenterPrint
-{
-  public:
-	virtual void ctor() = 0;
-	virtual void Clear(void) = 0;
-	virtual void ColorPrint(int r, int g, int b, int a, wchar_t* text) = 0;
-	virtual void ColorPrint(int r, int g, int b, int a, char* text) = 0;
-	virtual void Print(wchar_t* text) = 0;
-	virtual void Print(char* text) = 0;
-	virtual void SetTextColor(int r, int g, int b, int a) = 0;
-};
+		std::string name_str = "[NAME]";
+		int name_pos = str.find(name_str);
+		str.replace(name_pos, name_str.length(), msg.origin->ANSIColor + "[" + name + "]" + default_color);
 
-ICenterPrint* pInternalCenterPrint = NULL;
+		std::string level_str = "[LVL]";
+		int level_pos = str.find(level_str);
+		str.replace(level_pos, level_str.length(), levelColor + "[" + std::string(level_names[msg.level]) + "]" + default_color);
 
-enum class TextMsgPrintType_t
-{
-	HUD_PRINTNOTIFY = 1,
-	HUD_PRINTCONSOLE,
-	HUD_PRINTTALK,
-	HUD_PRINTCENTER
-};
+		out += str;
+	}
+	// print the string to the console - this is definitely bad i think
+	HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
+	auto ignored = WriteConsoleA(handle, out.c_str(), std::strlen(out.c_str()), nullptr, nullptr);
+	(void)ignored;
```

Comment:
I wrote this code, adapted from how spdlog itself does stuff, but I am not confident in it's functionality. If you think that it looks good, you could remove the comment saying that it is bad, or if it is actually bad, i would appreciate if you could improve it